src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs

```diff
@@ -54,17 +68,61 @@ public override void Dispose()
     private static readonly byte[] _jsonComma = [Convert.ToByte(',')];
     private static readonly byte[] _jsonClosingBracket = [Convert.ToByte(']')];
 
-    public override async Task ProcessAsync(ArraySegment<byte> data)
+    public override async Task ProcessAsync(ArraySegment<byte> data, CancellationToken cancellationToken)
     {
         IncrementBytesReceivedMetric(data.Count);
-        PipeReader request = PipeReader.Create(new ReadOnlySequence<byte>(data.Array!, data.Offset, data.Count));
+
+        IMemoryOwner<byte> memoryOwner = MemoryPool<byte>.Shared.Rent(data.Count);
+        data.AsSpan().CopyTo(memoryOwner.Memory.Span);
+        Memory<byte> memory = memoryOwner.Memory[..data.Count];
+
+        await _processChannel.Writer.WriteAsync(new ProcessRequest(memory, memoryOwner), cancellationToken);
+    }
+
+    public override async Task ReceiveLoopAsync(CancellationToken cancellationToken)
+    {
+        using AutoCancelTokenSource cts = cancellationToken.CreateChildTokenSource();
+
+        using ArrayPoolList<Task> allTasks = new(_processConcurrency + 1);
+        allTasks.Add(Task.Factory.StartNew(async () =>
+        {
+            try
+            {
+                await base.ReceiveLoopAsync(cts.Token);
+            }
+            finally
+            {
+                _processChannel.Writer.Complete();
+            }
+        }, TaskCreationOptions.LongRunning));
+
+        for (int i = 0; i < _processConcurrency; i++)
+        {
+            allTasks.Add(Task.Factory.StartNew(async () => await WorkerLoop(cts.Token), TaskCreationOptions.LongRunning));
+        }
+
+        await cts.WhenAllSucceed(allTasks);
+    }
+
+    private async Task WorkerLoop(CancellationToken cancellationToken)
+    {
+        await foreach (ProcessRequest request in _processChannel.Reader.ReadAllAsync(cancellationToken))
+        {
+            using var _ = request.BufferOwner;
+            await HandleRequest(request.Buffer, cancellationToken);
+        }
+    }
+
+    public async Task HandleRequest(Memory<byte> data, CancellationToken cancellationToken = default)
+    {
+        PipeReader request = PipeReader.Create(new ReadOnlySequence<byte>(data));
         int allResponsesSize = 0;
 
         await foreach (JsonRpcResult result in _jsonRpcProcessor.ProcessAsync(request, _jsonRpcContext))
```

Comment:
Should we cancel it here?
```suggestion
        await foreach (JsonRpcResult result in _jsonRpcProcessor.ProcessAsync(request, _jsonRpcContext).WithCancellation(cancellationToken))
```